core/src/main/scala/kafka/server/metadata/MetadataPartitions.scala

```diff
@@ -0,0 +1,282 @@
+      val m = new util.HashMap[Int, MetadataPartition](1)
+      changed.put(m, true)
+      m
+    } else if (changed.containsKey(prevPartitionMap)) {
+      prevPartitionMap
+    } else {
+      val m = new util.HashMap[Int, MetadataPartition](prevPartitionMap.size() + 1)
+      m.putAll(prevPartitionMap)
+      changed.put(m, true)
+      m
+    }
+    val prevPartition = newPartitionMap.put(partition.partitionIndex, partition)
+    if (partition.isReplicaFor(brokerId)) {
+      _localChanged.add(partition)
+    } else if (prevPartition != null && prevPartition.isReplicaFor(brokerId)) {
+      _localRemoved.add(prevPartition)
+    }
+    newNameMap.put(partition.topicName, newPartitionMap)
+  }
+
+  def remove(topicName: String, partitionId: Int): Unit = {
+    val prevPartitionMap = newNameMap.get(topicName)
+    if (prevPartitionMap != null) {
+      if (changed.containsKey(prevPartitionMap)) {
+        val prevPartition = prevPartitionMap.remove(partitionId)
+        if (prevPartition.isReplicaFor(brokerId)) {
+          _localRemoved.add(prevPartition)
+        }
+      } else {
+        Option(prevPartitionMap.get(partitionId)).foreach { prevPartition =>
+          if (prevPartition.isReplicaFor(brokerId)) {
+            _localRemoved.add(prevPartition)
+          }
+          val newPartitionMap = new util.HashMap[Int, MetadataPartition](prevPartitionMap.size() - 1)
+          prevPartitionMap.forEach(new BiConsumer[Int, MetadataPartition]() {
+            override def accept(key: Int, value: MetadataPartition): Unit =
+              if (!key.equals(partitionId)) {
+                newPartitionMap.put(key, value)
+              }
+          })
+          changed.put(newPartitionMap, true)
+          newNameMap.put(topicName, newPartitionMap)
+        }
+      }
+    }
+  }
+
+  def build(): MetadataPartitions = {
+    val result = new MetadataPartitions(newNameMap, newIdMap)
+    newNameMap = Collections.unmodifiableMap(newNameMap)
+    newIdMap = Collections.unmodifiableMap(newIdMap)
+    result
+  }
+
+  def localChanged(): Set[MetadataPartition] = _localChanged.asScala.toSet
+
+  def localRemoved(): Set[MetadataPartition] = _localRemoved.asScala.toSet
+}
+
+case class MetadataPartitions(private val nameMap: util.Map[String, util.Map[Int, MetadataPartition]],
+                              private val idMap: util.Map[Uuid, String]) {
+  def topicIdToName(uuid: Uuid): Option[String] = Option(idMap.get(uuid))
+
+  def copyNameMap(): util.Map[String, util.Map[Int, MetadataPartition]] = {
+    val copy = new util.HashMap[String, util.Map[Int, MetadataPartition]](nameMap.size())
+    copy.putAll(nameMap)
+    copy
+  }
+
+  def copyIdMap(): util.Map[Uuid, String] = {
+    val copy = new util.HashMap[Uuid, String](idMap.size())
+    copy.putAll(idMap)
+    copy
+  }
+
+  def allPartitions(): Iterator[MetadataPartition] = new AllPartitionsIterator(nameMap).asScala
+
+  def allTopicNames(): Set[String] = nameMap.keySet().asScala.toSet
```

Comment:
Another potentially dangerous copy here. I'm assuming this is unintended since I think we are trying to make the image immutable. We could fix this by using `collection.Set` as the return type and getting rid of the call to `toSet`.